plejd/mqtt.js

```diff
@@ -35,7 +35,7 @@ const getSettingsTopic = () => `plejd/settings`;
 const getDiscoveryPayload = device => ({
   schema: 'json',
   name: device.name,
-  unique_id: `light.plejd.${device.name.toLowerCase().replace(/ /g, '')}`,
+  unique_id: `light.plejd.${device.name.toLowerCase().replace(/ /g, '')}.${device.id}`,
```

Comment:
What kind if ID is device.id? Is it globally unique?

Shouldn't we remove the ${device.name} completley in that case?

(to avoid breaking all references when renaming the device in Plejd)

Reply:
I think i agree. I would probably expect (optimally) that the device is tracked as a physical unit. So renaming should optimally change device name, not id. 

If the id is indeed unique and programmed into the Plejd hardware that would mean removing a device in the Plejd app, moving it to a different socket and adding again, HA would still see it as the same device (which i think would be the golden standard).

Reply:
I don't think `{device.id}` is the solution to fix this in longterm though. In the Plejd- app and devices have a unique ID that goes under field `{device.serialNumber}`. 

Think the best solution is to replace `unique_id: light.plejd.${device.name.toLowerCase().replace(/ /g, '')}` with etc `unique_id: light.plejd.${device.serialNumber.toLowerCase().replace(/ /g, '')}` to be sure that the device always stays. 

Then I'm not sure what it will be like for scenes. So it needs to be investigated further.

Reply:
Checked the scenes, they also has an unique identifier under field `{device.serialNumber}`

Reply:
@xpathor This looks promizing. Do you have time to test and update the pull request?

Reply:
FYI I tried that change (using serialNumber only) locally, which generally works, but the DIM-02 units now don't get properly recognized (only one of them can be used). I suspect they report the same device ID, so we get a similar issue as with the name clash in the original bug.

I didn't have time yet to see if there is something straight forward we could do. Was gonna have a look at the plejd API response contents, and then think about it some more.

Reply:
My solution to the issues with DIM-02, at the time when I wrote the addon, was to use the name. That isn't a good identifier but was what I found most usable from my perspective at the moment. I agree with you that this needs to be addressed.

I'll dump the Plejd API response and have a look myself aswell to see what is usable.

Reply:
```
"inputSettings": [
        {
          "deviceId": "C1FFF1D9AEAA",
          "siteId": "329dad16-f40b-0000-0000-a2e194f92976",
          "input": 1,
          "buttonType": "PushButton",
          "dimSpeed": 2,
          "doubleSidedDirectionButton": false,
          "createdAt": "2020-12-19T13:55:49.920Z",
          "updatedAt": "2020-12-19T13:55:49.920Z",
          "ACL": {},
          "objectId": "A21MZaEtZZ",
          "__type": "Object",
          "className": "PlejdDeviceInputSetting"
        },
        {
          "deviceId": "C1FFF1D9AEAA",
          "siteId": "329dad16-f40b-0000-0000-a2e194f92976",
          "input": 0,
          "buttonType": "Other",
          "dimSpeed": -1,
          "doubleSidedDirectionButton": false,
          "createdAt": "2020-12-19T13:55:49.920Z",
          "updatedAt": "2021-01-15T17:19:31.357Z",
          "ACL": {},
          "objectId": "CJ8Rk8ZZ7n",
          "__type": "Object",
          "className": "PlejdDeviceInputSetting"
        }
      ],
```

DeviceId is the hardware device, as you said.
Is it possible to combine like `deviceId:input`? Or use objectId (probably not hardware bound, but installation bound, will probably change identity if you reset it, but does it really matter?)

Reply:
Deleted my last comment but I changed my mind so I'm posting it again

Maybe instead of parsing .devices we should parse the switches in .inputSettings (from getSiteById)?

Then you can have it the unqiue id as deviceId + input. (Maybe something like C1FFF1D9AEAA:0 and C1FFF1D9AEAA:1)

```
"inputSettings": [
        {
          "deviceId": "C1FFF1D9AEAA",
          "siteId": "329dad16-f40b-0000-0000-a2e194f92976",
          "input": 1,
          "buttonType": "PushButton",
          "dimSpeed": 2,
          "doubleSidedDirectionButton": false,
          "createdAt": "2020-12-19T13:55:49.920Z",
          "updatedAt": "2020-12-19T13:55:49.920Z",
          "ACL": {},
          "objectId": "A21MZaEtZZ",
          "__type": "Object",
          "className": "PlejdDeviceInputSetting"
        },
        {
          "deviceId": "C1FFF1D9AEAA",
          "siteId": "329dad16-f40b-0000-0000-a2e194f92976",
          "input": 0,
          "buttonType": "Other",
          "dimSpeed": -1,
          "doubleSidedDirectionButton": false,
          "createdAt": "2020-12-19T13:55:49.920Z",
          "updatedAt": "2021-01-15T17:19:31.357Z",
          "ACL": {},
          "objectId": "CJ8Rk8ZZ7n",
          "__type": "Object",
          "className": "PlejdDeviceInputSetting"
        }
      ],

```

Reply:
I don't know much about objectId vs deviceId, _but_ what I have noticed is that we have one more case we don't support: devices where the load is on another Plejd device.

So: I have a few 'rogue' dimmers that controls other Plejd devices only. Ex: Kitchen dimmer [dev id A] => [no load connected] => BLE => Kitchen pendulum light [dev id B].

Kitchen pendulum light is picked up correctly, the dimmer is not. Rotating the dimmer (physically) gives "unknown device" warnings in the addon. The slave device (Kitchen pendulum light) will however still be updated correctly by the next BLE command related to that device id.

In the API response, that specific device (id 116) only shows up in `outputGroups` and `inputAddress` lookup tables.

```
  "outputGroups": {
    ...
    "E8153A3CBF90": { "0": [10, 22, 85, 86, 87, 88, 116] },
    "FCDA8C56CBDD": { "0": [10, 22, 81, 82, 83, 84, 116] },
     ...
  },
  "inputAddress": {
    ...
    "E352510EC81D": { "0": 12, "1": 12, "2": 116 },
    ...
  },
```

The input address `E352510EC81D` is the physical dimmer I changed, The output group id:s `E8153A3CBF90` and `FCDA8C56CBDD` are the two kitchen pendulums each having its own Plejd device that only acts as slave to others.

Reply:
Btw: `deviceId` is what is persisted in the code base (in device registry) as `serialNumber`, also used to lookup API device based on discovered BLE device.

Reply:
> So: I have a few 'rogue' dimmers that controls other Plejd devices only. Ex: Kitchen dimmer [dev id A] => [no load connected] => BLE => Kitchen pendulum light [dev id B].

The 'rogue" devices should (IMO) not be injected to Home Assistant, as they have no meaning. They cannot be controlled anyway. In the Plejd app, these devices are hidden from the Ui.

This information can be found in the API, but a bit complex to find:

Since DIM-02 has two loads (and indexes), the `objectId `needs to be used.

Sadly, one could have hoped for `devices[objectId].hiddenFromRoomList`, but it will not change when you remove the load.

Instead, `outputSettings[deviceParseId=objectId].predefinedLoad.loadType == "No load"` indicates that the room should be hidden from HA.

`outputSettings[deviceParseId=objectId].deviceId` provides the `deviceId`
`outputSettings[deviceParseId=objectId].output` tells for a DIM-02 if it is `output "0"` or `"1"`
`outputAddress[deviceId[output]] `gives the `index`

```
  "outputSettings": [
    {
      "deviceId": "D670E6FE619D",
      "siteId": "b9079ce2-4ca8-4a16-863a-16944520292b",
      "output": 0,
      "predefinedLoad": {
        "loadType": "Halogen Dim02",
        },
      "deviceParseId": "xJq8dq5K3C",
    },
    {
      "deviceId": "D670E6FE619D",
      "siteId": "b9079ce2-4ca8-4a16-863a-16944520292b",
      "output": 1,
      "predefinedLoad": {
        "loadType": "No load",
        },
      "deviceParseId": "6kKFTzP1YD",
    }
```
```
  "devices": [
    {
      "deviceId": "D670E6FE619D",
      "siteId": "b9079ce2-4ca8-4a16-863a-16944520292b",
      "title": "Dimmer_2_1",
      "hiddenFromRoomList": false,
      "hiddenFromIntegrations": false,
      "objectId": "xJq8dq5K3C",
    },    {
      "deviceId": "D670E6FE619D",
      "siteId": "b9079ce2-4ca8-4a16-863a-16944520292b",
      "title": "Dimmer_2_2",
      "hiddenFromRoomList": false,
      "hiddenFromIntegrations": false,
      "objectId": "6kKFTzP1YD",
    }
```
```
  "outputAddress": {
    "D670E6FE619D": {
      "0": 15,
      "1": 16
    }
```